File: src/fast_edit/tag_range.rs
use std::fs;
use std::path::Path;

use crate::error::{EditError, EditResult};

#[derive(serde::Serialize)]
pub struct TagRangeResult {
    pub start: usize,
    pub end: usize,
    pub kind: String,
}

pub fn op_tag_range(filepath: &str, line: usize) -> EditResult<TagRangeResult> {
    let content = fs::read_to_string(Path::new(filepath))
        .map_err(|e| EditError::read_path(filepath, e))?;
    let lines: Vec<&str> = content.split('\n').collect();
    if line < 1 || line > lines.len() {
        return Err(EditError::invalid_arg(format!("目标行 {} 超出文件范围 (1..{})", line, lines.len())));
    }

    let mut stack: Vec<(String, usize)> = Vec::new();
    let mut open_ranges: Vec<(String, usize, usize)> = Vec::new();

    for (idx, raw_line) in lines.iter().enumerate() {
        let line_no = idx + 1;
        let mut cursor = 0usize;
        let chars: Vec<char> = raw_line.chars().collect();
        while cursor < chars.len() {
            if chars[cursor] == '<' {
                let mut j = cursor + 1;
                if j < chars.len() && chars[j] == '!' {
                    while j < chars.len() && chars[j] != '>' {
                        j += 1;
                    }
                    cursor = j.saturating_add(1);
                    continue;
                }
                if j < chars.len() && chars[j] == '/' {
                    j += 1;
                    let mut tag = String::new();
                    while j < chars.len() {
                        let ch = chars[j];
                        if ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' || ch == ':' {
                            tag.push(ch);
                            j += 1;
                        } else {
                            break;
                        }
                    }
                    while j < chars.len() && chars[j] != '>' {
                        j += 1;
                    }
                    if let Some((open_tag, open_line)) = stack.pop()
                        && open_tag == tag
                    {
                        open_ranges.push((open_tag, open_line, line_no));
                    }
                    cursor = j.saturating_add(1);
                    continue;
                }
                let mut tag = String::new();
                while j < chars.len() {
                    let ch = chars[j];
                    if ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' || ch == ':' {
                        tag.push(ch);
                        j += 1;
                    } else {
                        break;
                    }
                }
                if !tag.is_empty() {
                    stack.push((tag, line_no));
                }
                while j < chars.len() && chars[j] != '>' {
                    j += 1;
                }
                cursor = j.saturating_add(1);
                continue;
            }
            cursor += 1;
        }
    }

    for (kind, start, end) in open_ranges {
        if start <= line && line <= end {
            return Ok(TagRangeResult { start, end, kind });
        }
    }

    Err(EditError::invalid_arg(format!(
        "第 {} 行不在任何可配对 tag 范围内",
        line
    )))
}

File: internal/edit/tag_range.go
package edit

import (
	"fmt"
	"strings"
)

func TagRange(path string, line int) (TagRangeResult, error) {
	content, err := ReadText(path)
	if err != nil {
		return TagRangeResult{}, ReadPath(path, err)
	}
	lines := strings.Split(content, "\n")
	if line < 1 || line > len(lines) {
		return TagRangeResult{}, InvalidArg(fmt.Sprintf("目标行 %d 超出文件范围 (1..%d)", line, len(lines)))
	}

	type tagEntry struct {
		name string
		line int
	}
	var stack []tagEntry
	var openRanges []TagRangeResult
	voidElements := map[string]struct{}{
		"area": {}, "base": {}, "br": {}, "col": {}, "embed": {}, "hr": {}, "img": {},
		"input": {}, "link": {}, "meta": {}, "param": {}, "source": {}, "track": {}, "wbr": {},
	}

	for idx, rawLine := range lines {
		lineNo := idx + 1
		chars := []byte(rawLine)
		for cursor := 0; cursor < len(chars); cursor++ {
			if chars[cursor] != '<' {
				continue
			}
			j := cursor + 1
			if j < len(chars) && chars[j] == '!' {
				for j < len(chars) && chars[j] != '>' {
					j++
				}
				cursor = j
				continue
			}
			if j < len(chars) && chars[j] == '/' {
				j++
				tag := readTagName(chars, &j)
				for j < len(chars) && chars[j] != '>' {
					j++
				}
				if len(stack) > 0 {
					last := stack[len(stack)-1]
					if last.name == tag {
						stack = stack[:len(stack)-1]
						openRanges = append(openRanges, TagRangeResult{Start: last.line, End: lineNo, Kind: last.name, Tag: last.name})
					}
				}
				cursor = j
				continue
			}
			tag := readTagName(chars, &j)
			if tag != "" {
				trimmed := strings.TrimRight(string(chars[cursor:]), " \t")
				if strings.HasSuffix(trimmed, "/>") {
					for j < len(chars) && chars[j] != '>' {
						j++
					}
					cursor = j
					continue
				}
				if _, ok := voidElements[strings.ToLower(tag)]; !ok {
					stack = append(stack, tagEntry{name: strings.ToLower(tag), line: lineNo})
				}
			}
			for j < len(chars) && chars[j] != '>' {
				j++
			}
			cursor = j
		}
	}

	for _, rg := range openRanges {
		if rg.Start <= line && line <= rg.End {
			return rg, nil
		}
	}
	return TagRangeResult{}, InvalidArg(fmt.Sprintf("第 %d 行不在任何可配对 tag 范围内", line))
}

func readTagName(chars []byte, idx *int) string {
	j := *idx
	var tag strings.Builder
	for j < len(chars) {
		ch := chars[j]
		if (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') || ch == '-' || ch == '_' || ch == ':' {
			tag.WriteByte(ch)
			j++
			continue
		}
		break
	}
	*idx = j
	return tag.String()
}
